### ISSUE TITLE
Possible bug in HASH_ADD_KEYPTR_INORDER causing segfault

### ISSUE THREAD

**Issue by IronBug** [troydhanson/uthash#99]
Greetings.
I've been using uthash for years but recently it seems I found a bug in it. 
Bug is in pointer keys with sorting function, i.e. in HASH_ADD_KEYPTR_INORDER. And it causes segfault.
It seems strange, but it really looks like a bug. I read documentation some times and I guess I used it properly, but it failed. 

I made a small test case to show the bug. 
I also sketched a patch (only for GCC, since I don't have other compilers at hand, but it's easily extendable for other declarations).
I uploaded the test case and patch here:
https://github.com/IronBug/uthash_test

Sincerely, Yana A. Kireyonok aka Iron Bug

**Comment by Quuxplusone** [comment]
Could you post the patch also as a pull request against this repo?  I'm inclined to investigate deeply before accepting anything along those lines, but submitting it as a PR would provide a better UI for people looking at the patch. Also, it would give you a chance to add your test case to the tests/ directory.

I see two perhaps functional changes in that patch:

(1) The current code trusts that `DECLTYPE(head) ELMT_FROM_HH((head)->hh.tbl, _hs_iter)` evaluates to a real object. This might break some aliasing assumptions on optimizing compilers, maybe. In your patched version, you eliminate all that macro magic by dealing in `_el_iter` directly.

(2) In your patched version, you assign `(add)->hh.next = _el_iter;` instead of `(add)->hh.next = _hs_iter;`. I do believe this is a bug in the existing code, and I'm amazed the existing tests didn't catch it! However, I have not yet figured out why the whole patch doesn't boil down to a one-liner:

```
diff --git a/src/uthash.h b/src/uthash.h
index 45d1f9f..5dd64f9 100644
--- a/src/uthash.h
+++ b/src/uthash.h
@@ -249,7 +249,7 @@ do {
         break;                                                                   \
     } while ((_hs_iter = _hs_iter->next));                                       \
     if (_hs_iter) {                                                              \
-      (add)->hh.next = _hs_iter;                                                 \
+      (add)->hh.next = ELMT_FROM_HH((head)->hh.tbl, _hs_iter);                   \
       if (((add)->hh.prev = _hs_iter->prev)) {                                   \
         HH_FROM_ELMT((head)->hh.tbl, _hs_iter->prev)->next = (add);              \
       } else {                                                                   \
```

**Comment by Quuxplusone** [comment]
...And the existing tests didn't catch it because there are no tests. :D
```
Arthurs-MacBook-Pro-2:uthash ajo$ git grep INORDER tests
Arthurs-MacBook-Pro-2:uthash ajo$ 
```
So yours can be the first!
But I'd still like to either investigate-and-understand, or be-told-and-understand, why the one-liner above isn't sufficient to fix the segfault. (It's not. I tried.)

**Comment by IronBug** [comment]
I did not meant it as a ready-to-apply patch. It was just a fast hack to fix the problem, actually. I made it for my own need to get rid of segfault. So it may be out of conception. Because I did not come with the idea how to put it into the existing offsets conception, So I made the simpliest change that came to my mind to fix it and just worked with items in list directly.
I made the patch just to show where the bug hides. It's in addressing, actually. The offset changes in a wrong way.

**Comment by IronBug** [comment]
Actually, if this could help somehow, the bug is in line 248:
      if (cmpfcn(DECLTYPE(head) ELMT_FROM_HH((head)->hh.tbl, _hs_iter), add) > 0) \
ecactly in the first parameter to cmpfcn. 
On the first iteration the offset is correct, but the second iteration offset is wrong, so function gets wrong pointer and fails. In my case the offset passed to function was 24 bytes less than it should be. Maybe this could clear the case a little.

**Comment by Quuxplusone** [comment]
I have a tentative fix to the header file here: https://github.com/Quuxplusone/uthash/commit/61f121ec035b14b222e3fb5a03c62a58e844a1ce but I'll add some tests before actually submitting a PR.

**Comment by IronBug** [comment]
On my side this fix works fine, I compiled the program that I worked on when discovered a bug and now it works fine. It uses thousands of sorted items with two levels of pointer keys, with different sort orders. And now there's no problem, So I assume the bug is fixed.
Thanks.

**Comment by Quuxplusone** [comment]
@IronBug, do you think you could provide some code besides the test case you already provided, e.g. different sort orders and so on, for me to incorporate into the uthash test suite?

It would not surprise me at all if you're the first person ever to attempt using the `INORDER` macros. These macros are relatively new; they were only added in 650214c076dc04d2b1d9dc1a25dc082032f7782e (June 2016) based on work by @thiloschulz. I would have assumed he'd been using them, except that AFAICT this bug would have prevented using them on any hash table of more than a single element, so he couldn't have been using them *too* much. Most likely, you're in completely untested territory... which I know is a terrible thing to hear from a library maintainer. :)

**Comment by thiloschulz** [comment]
It is true, after writing these macros a design decision I made in my software obliviated the need for the INORDER macros so I did not run into this.

test87.c is a testcase I wrote and it ran fine.

**Comment by Quuxplusone** [comment]
My apologies, test87.c does indeed exist. I somehow missed that when I did my grep yesterday.

**Comment by Quuxplusone** [comment]
Fixed by #100 (13723b1c0d16e2a80e0f6bf0175aaadcd6a6c8f5).
